File: src/my_cpp_pkg/src/number_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include <string>

class NumberPublisher : public rclcpp::Node
{
public:
    NumberPublisher() : Node("number_publisher")
    {
        __publisher = this->create_publisher<example_interfaces::msg::Int64>("number", 10);
        __timer = this->create_wall_timer(std::chrono::milliseconds(500),
                                          std::bind(&NumberPublisher::publishNews, this));
        RCLCPP_INFO(this->get_logger(), "");
    }

private:
    rclcpp::TimerBase::SharedPtr __timer;
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr __publisher;

    void publishNews()
    {
        auto msg = example_interfaces::msg::Int64();
        msg.data = 42;
        __publisher->publish(msg);
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberPublisher>(); // Cpp nodes are created as shared pointer
    rclcpp::spin(node);
    rclcpp::shutdown();

    return 0;
}

File: src/my_py_pkg/my_py_pkg/number_publisher.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.msg import (
    Int64,
)  # need to add this dependency in package.xml too


class NumberPublisher(Node):
    def __init__(self):
        super().__init__("number_publisher")

        self.publisher = self.create_publisher(Int64, "number", 10)
        self.timer = self.create_timer(0.5, self.publish_number)
        self.get_logger().info("Number publisher node has started!")

    def publish_number(self):
        msg = Int64()
        msg.data = 42
        self.publisher.publish(msg)


def main(args=None):
    rclpy.init(args=args)  # initialize ROS
    node = NumberPublisher()  # create node
    rclpy.spin(node)  # allows the node to run indefintely
    rclpy.shutdown()  # shutdown node


if __name__ == "__main__":
    main()
